src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -47,405 +49,429 @@
-      grpc_channel_args_destroy(args->args);
-      gpr_ref(&connection_state->refs);
-      GRPC_CHTTP2_REF_TRANSPORT((grpc_chttp2_transport*)transport,
-                                "receive settings timeout");
-      GRPC_CLOSURE_INIT(&connection_state->on_timeout, on_timeout,
-                        connection_state, grpc_schedule_on_exec_ctx);
-      grpc_timer_init(&connection_state->timer, connection_state->deadline,
-                      &connection_state->on_timeout);
-    } else {
+void Chttp2ServerListener::ConnectionState::OnHandshakeDone(void* arg,
+                                                            grpc_error* error) {
+  auto* args = static_cast<HandshakerArgs*>(arg);
+  ConnectionState* self = static_cast<ConnectionState*>(args->user_data);
+  {
+    MutexLock lock(&self->listener_->mu_);
+    grpc_resource_user* resource_user =
+        grpc_server_get_default_resource_user(self->listener_->server_);
+    if (error != GRPC_ERROR_NONE || self->listener_->shutdown_) {
+      const char* error_str = grpc_error_string(error);
+      gpr_log(GPR_DEBUG, "Handshaking failed: %s", error_str);
+      grpc_resource_user* resource_user =
+          grpc_server_get_default_resource_user(self->listener_->server_);
       if (resource_user != nullptr) {
         grpc_resource_user_free(resource_user,
                                 GRPC_RESOURCE_QUOTA_CHANNEL_SIZE);
       }
+      if (error == GRPC_ERROR_NONE && args->endpoint != nullptr) {
+        // We were shut down after handshaking completed successfully, so
+        // destroy the endpoint here.
+        // TODO(ctiller): It is currently necessary to shutdown endpoints
+        // before destroying them, even if we know that there are no
+        // pending read/write callbacks.  This should be fixed, at which
+        // point this can be removed.
+        grpc_endpoint_shutdown(args->endpoint, GRPC_ERROR_NONE);
+        grpc_endpoint_destroy(args->endpoint);
+        grpc_channel_args_destroy(args->args);
+        grpc_slice_buffer_destroy_internal(args->read_buffer);
+        gpr_free(args->read_buffer);
+      }
+    } else {
+      // If the handshaking succeeded but there is no endpoint, then the
+      // handshaker may have handed off the connection to some external
+      // code, so we can just clean up here without creating a transport.
+      if (args->endpoint != nullptr) {
+        grpc_transport* transport = grpc_create_chttp2_transport(
+            args->args, args->endpoint, false, resource_user);
+        grpc_server_setup_transport(
+            self->listener_->server_, transport, self->accepting_pollset_,
+            args->args, grpc_chttp2_transport_get_socket_node(transport),
+            resource_user);
+        // Use notify_on_receive_settings callback to enforce the
+        // handshake deadline.
+        self->transport_ = reinterpret_cast<grpc_chttp2_transport*>(transport);
```

Comment:
To me reinterpret_cast is a code smell but I understand why it works here. Can I suggest a comment above saying that this reinterpret_cast is safe because grpc_chttp2_transport is currently a C-style extension of grpc_transport and thus morally equivalent to a static_cast to a derived class?